operator/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -112,390 +107,319 @@ type RedpandaReconciler struct {
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *RedpandaReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
-	if err := registerHelmReferencedIndex(ctx, mgr, "statefulset", &appsv1.StatefulSet{}); err != nil {
-		return err
-	}
-	if err := registerHelmReferencedIndex(ctx, mgr, "deployment", &appsv1.Deployment{}); err != nil {
-		return err
-	}
+	builder := ctrl.NewControllerManagedBy(mgr)
 
-	helmManagedComponentPredicate, err := predicate.LabelSelectorPredicate(
-		metav1.LabelSelector{
-			MatchExpressions: []metav1.LabelSelectorRequirement{{
-				Key:      "app.kubernetes.io/name", // look for only redpanda or console pods
-				Operator: metav1.LabelSelectorOpIn,
-				Values:   []string{"redpanda", "console", "connectors"},
-			}, {
-				Key:      "app.kubernetes.io/instance", // make sure we have a cluster name
-				Operator: metav1.LabelSelectorOpExists,
-			}, {
-				Key:      "batch.kubernetes.io/job-name", // filter out the job pods since they also have name=redpanda
-				Operator: metav1.LabelSelectorOpDoesNotExist,
-			}},
-		},
-	)
-	if err != nil {
+	if err := r.LifecycleClient.WatchResources(builder, &redpandav1alpha2.Redpanda{}); err != nil {
 		return err
 	}
 
-	managedWatchOption := builder.WithPredicates(helmManagedComponentPredicate)
-
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&redpandav1alpha2.Redpanda{}).
-		Watches(&appsv1.StatefulSet{}, enqueueClusterFromHelmManagedObject(), managedWatchOption).
-		Watches(&appsv1.Deployment{}, enqueueClusterFromHelmManagedObject(), managedWatchOption).
-		Complete(r)
+	return builder.Complete(r)
 }
 
-func (r *RedpandaReconciler) Reconcile(c context.Context, req ctrl.Request) (ctrl.Result, error) {
-	ctx, done := context.WithCancel(c)
-	defer done()
-
-	start := time.Now()
+func (r *RedpandaReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx).WithName("RedpandaReconciler.Reconcile")
 
-	defer func() {
-		durationMsg := fmt.Sprintf("reconciliation finished in %s", time.Since(start).String())
-		log.V(logger.TraceLevel).Info(durationMsg)
-	}()
-
-	log.V(logger.TraceLevel).Info("Starting reconcile loop")
+	defer timing.Execution(ctx).Stop("reconciling redpanda cluster")
 
 	rp := &redpandav1alpha2.Redpanda{}
 	if err := r.Client.Get(ctx, req.NamespacedName, rp); err != nil {
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
+	rp.ManagedFields = nil // nil out our managed fields
```

Comment:
IMO this should be done immediately before any writes. This will get forgotten as is.